### PR TITLE
Update domain status without WlsDomainConfig

### PR DESCRIPTION
**Pull request by russgold**
Addresses an issue with detection and reporting of the ReplicasTooHigh failure (OWLS-93692). Now confirmed by an integration test.

Ultimately, this change addresses a longstanding design error that wasn't so much of an issue before we added the above feature: much of the status computation depended on the WlsDomainConfig, which is only available during a make-right, and not during the more frequent scheduled status updating. As a result, we were bypassing that processing. With ReplicasTooHigh, though, the result was that the conditions, which were still being computed, were computed incorrectly, since we couldn't tell that we were missing the cluster limits. 

This change segregates the status computation into two parts: the basic server and cluster statuses actually can only be determined from the WlsDomainConfig, and the rest, which can then use the information recorded in the status itself.

One consequence of this change is that server and cluster status updates should happen more quickly, since they no longer need to wait for the next make-right.

### CODE REVIEW

**Review by lennyphan — APPROVED** [review]
LGTM

**Review by doxiao — COMMENTED** [review]
LGTM overall. Had a couple of comments.

File: operator/src/main/java/oracle/kubernetes/operator/DomainStatusUpdater.java [hunk 1]
```diff
@@ -569,15 +571,20 @@ private boolean haveServerData() {
...
-        private final ClusterCheck[] clusterChecks = createClusterChecks();
+        private final DomainStatus status;
+        private final ClusterCheck[] clusterChecks;
         private final List<DomainCondition> conditions = new ArrayList<>();
 
-        void apply(DomainStatus status) {
+        void apply() {
           status.removeConditionsMatching(c -> c.hasType(Failed) && ReplicasTooHigh.name().equals(c.getReason()));
-          conditions.forEach(status::addCondition);
+          conditions.forEach(newCondition -> addCondition(status, newCondition));
```
[the remark this answers is not in the document]
Comment (russgold) [reply operator/src/main/java/oracle/kubernetes/operator/DomainStatusUpdater.java:580]: addCondition will ignore conditions that are already present, and add only changed ones, removing those that mismatch on status.

File: operator/src/main/java/oracle/kubernetes/operator/DomainStatusUpdater.java [hunk 2]
```diff
@@ -590,16 +597,16 @@ void apply(DomainStatus status) {
...
-                .orElse(Collections.emptyList());
+        private ClusterCheck createFrom(ClusterStatus clusterStatus) {
+          return new ClusterCheck(status, clusterStatus);
         }
 
-        public Conditions() {
+        public Conditions(DomainStatus status) {
+          this.status = status != null ? status : new DomainStatus();
+          this.clusterChecks = createClusterChecks();
           conditions.add(new DomainCondition(Completed).withStatus(isProcessingCompleted()));
```
[the remark this answers is not in the document]
Comment (russgold) [reply operator/src/main/java/oracle/kubernetes/operator/DomainStatusUpdater.java:610]: It's not necessary, as noted above. The logic is in the DomainStatus.

File: operator/src/main/java/oracle/kubernetes/operator/DomainStatusUpdater.java [hunk 3]
```diff
@@ -703,14 +719,81 @@ private String createFailureMessage() {
...
+        getDomainConfig()
+              .map(c -> new DomainStatusFactory(getDomain(), c, this::isStartedServer))
+              .ifPresent(f -> f.setStatusDetails(status));
       }
 
-      @Nonnull
-      private String getAdminServerName() {
-        return getDomainConfig().map(WlsDomainConfig::getAdminServerName).orElse("");
+      private void updateStatusDetails(DomainStatus status) {
+        new StatusDetailsUpdate(status).updateStatusDetails(status);
```
[the remark this answers is not in the document]
Comment (russgold) [reply operator/src/main/java/oracle/kubernetes/operator/DomainStatusUpdater.java:728]: Good catch. I missed this while refactoring. I will update.

**Review by doxiao — APPROVED** [review]

**Review by doxiao — APPROVED** [review]
LGTM

**Pull request merged**